File: 0075-sort-colors/0075-sort-colors.java
class Solution {
    public void sortColors(int[] nums) {
        int low = 0;           // pointer for 0's boundary
        int mid = 0;           // pointer to traverse the array
        int high = nums.length - 1; // pointer for 2's boundary

        // Traverse the array till mid passes high
        while (mid <= high) {
            if (nums[mid] == 0) {
                // Swap nums[mid] and nums[low]
                int temp = nums[low];
                nums[low] = nums[mid];
                nums[mid] = temp;
                low++;  // increment low as 0 is placed at the front
                mid++;  // increment mid to move to next element
            } else if (nums[mid] == 1) {
                // Just move mid ahead, 1 is in correct position
                mid++;
            } else {
                // nums[mid] == 2
                // Swap nums[mid] and nums[high]
                int temp = nums[mid];
                nums[mid] = nums[high];
                nums[high] = temp;
                high--; // decrement high as 2 is placed at the end
                // don't increment mid here
            }
        }
    }
}

File: 0075-sort-colors/0075-sort-colors.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function(nums) {
    let n = nums.length;
    let low = 0;
    let mid = 0;
    let high = n - 1;

    while(mid <= high) {
        if(nums[mid] == 0) {
            swap(nums, low, mid);
            low++;
            mid++;
        } else if(nums[mid] == 1) {
            mid++;
        } else {    
            swap(nums, mid, high);
            high--;
        }
    }
};

function swap(nums, i, j) {
    let temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
}